Monika-After-Story/MonikaModDev
# More Consistent Greeting Rules

#3531 (partial)

(based to `content` so content stuff can use this right away)

# Key Features + Changes

## Greetings

* `selectGreeting` has been entirely redone. The full alg is explained below.
* new function `_filterGreeting`, which is used to evaluate a single greeting for the new greeting rules. (See below)
* `random` is no longer checked. There is an update script to clear random from all Greeting objects.

### New Greeting Alg

1. Each greeting is evaluted using `_filterGreeting`
2. If it passes the filter, greeting priority is compared to current priority:
    1. If same priority as current, we add it to the available pool.
    2. If lower in value than current (higher priority), we drop the pool and create a new pool with the greeting and its priority.
3. If we have nothing in the pool, we return None.
4. otherwise, we return a random selection from the pool.

### Filter Rules

Here are the new rules that apply to greetings:
* affection - `aff_range` is now respected by greeting selection alg
* conditional - `conditoinal` is now respected by greeting selection alg. This should be used sparingly. It is much easier to control greetings with `unlocked` than conditional.
* priority - provided by `MASPriorityRule`. Events with a lower integer value are considered **higher priority**. 
* random chance - provided by `MASGreetingRule`. Only checked if the greeting has this rule. 
* repeats - `MASSelectiveRepeatRule` and `MASNumericalRepeatRule` are still respected, but only checked if the greeting has either/both rules.
* type - if type is provided, greetings **must** have this in their `category` property. If type is **not** provided, greetings **must** have a **None** for their `category` property. `persistent._mas_greeting_type` should **only be a single type**. Types are no longer checked via set intersections.
* unlocked - greetings **MUST** be `unlocked` for them to be eligible for selection. This is the sole driver of whether or not greetings can be selected.

And the order that they are evaluted in:
1. Priority - compared to the current priority, which is always equal to the lowest one that is in the selection pool.
2. Type 
3. Unlocked
4. Affection
5. `MASSelectiveRepeatRule`
6. `MASNumericalRepeatRule`
7. `MASGreetingRule` - random chance
8. conditional

### Greeting Priorities
* the default priority for events is 500. 
* non-special events should be between 50-100.
* special events should have 10-50 
* debug greetings should have negative priority.

## Event
* `rules` is no longer saved to persistent. Data in this property is now saved as None in the tuples, nor is this verified on startup. as a result of this, index 14 is now a "hole" in the data model. A data migration system (#3547) is planned to address this issue.
* `_checkRepeatRule` is now DEPRECATED. This should be removed once farewells is updated.
* `checkRepeatRules` is now DEPRECATED. This should be removed once farewells is updated.
* `_checkGreetingRule` has been removed. Use `MASGreetingRule.evaluate_rule` instead.
* `checkGreetingRules` has been removed. There was no point in only checking `MASGreetingRules` for events.

## Event Rules

### MASSelectiveRepeatRule
* `evaluate_rule` now has support for a `defval` (default value) param. This is returned if the EV does not have this rule.

### MASNumericalRepeatRule
* `evaluate_rule` now has support for a `defval` (default value) param. This is returned if the EV does not have this rule.

### MASGreetingRule
* `evaluate_rule` now has support for a `defval` (default value) param. This is returned if the EV does not have this rule.
* Added `setup_label` as a rule param, this label is called after the greeting has been selected but before `ch30_post_greeting_check`. This is designed for greetings that need to do setup like a potential crashed greeting.
* `random_chance` no longer fails the rule if 0. Instead it returns `defval` as if the rule did not exist.

### MASPriorityRule
* contains an integer, which acts as a priority. The actual implementation and meaning behind this integer should be determined by the implementing module.

## Docking Station
* `selectReturnHomeGreeting` is no longer custom, it calls `selectGreeting` using the given greeting type. If `selectGreeting` returns None, this function returns the standard returned home greeting object.

## Affection
* removed event/greeting code that used affection programming points. Greetings should be using `aff_range` now.

## Delayed Action
* Delayed Action ID 1 (ourreality unlock) has been removed. Using `aff_range` has negated the point of this Delayed Action. ID 1 is retired and should not be used in any future DelayedActions.

## Dev
* new dev event `SAMPLE GRE`, which lets you test greetings with a sample size. Designed for looking at the random chances greetings can be selected by the new system.

## Specific Greeting Changes

In general, all greetings have been updated to use `aff_range`. Most greetings have default priority as well. 
Ones that are special are listed here.

* `opendoor` - now has a priority of 50
* `monikaroom_will_change` - has a priority of 10. This should probably be the sole event with this priority, exception of potentially final farewell if it becomes a greeting instead.
* `timeconcern` and `timeconcern_day` - both of these have a priority of 70.
* `hairdown` - now has a priority of 45.
* `tears` - now has a priority of 15
* `upset` - no longer has a random chance
* `distressed` - no longer has a random chance
* `ourreality` - unlocked by default. Has a priority of 40.
* `trick_or_treat_back` - removed `TYPE_GO_SOMEWHERE` from this, as types are now checked as individuals, and not as sets.

## Update scripts

* all greetings `random` properties are set to False
* `monika_rain` is now unlocked if you have seen the label. This might have been locked accidentally via affection programming points.
* islands greeting is unlocked if you have not seen it yet. It now triggers if affection is appropriate.

# Testing
* regression test all greetings

Running the sampler is the easy way to tell if greetings are being appropriately selected over a large number of runs. The gist is to:
1. Set the conditions/properties of the greeting so that it can be triggered (unlocking/changing affection/...)
2. Run the greeting sampler. There are 2 prompts that let you know when to set `use_type` and `check_time`. These vars let you feed a specific greeting type or datetime to the `selectGreeting` function.
3. Evaluate results in `basedir/gre_sample`. For easy viewing, add a `csv` ext to the end of the file and import into spreadsheet program.

**NOTE:** the sampler is limited to 10k runs. Above that would probably make the game freeze for longer than 5 minutes.

For certain special greetings, it might be worthwhile testing them out normally:
* docking station greetings 
    * only need to test the main one and trick or treat. The changes would only affect those 2 docking station greetings anyway.
* crashed greeting - just verify that it still is shown on a crash. This will be formally replaced in #3435.
* forced greetings - the o31 ones, although I'm nearly certain that there won't be issues here. forced greetings take their own path around the greeting selection functon anyway.
